### PR TITLE
This PR completes the Milestone Migrate Project Explorer

### PR DESCRIPTION
**Pull request by timbittner**
All functions of the original SWT ProjectExplorer are now available in the FX ProjectExplorer.

New features since last PR:

- DoubleClickListener to open folder nodes and special files (file endings defined by other plugins) https://github.com/Jooshua/e4CompareFramework-team2/issues/11
- DropTarget to quickly import files from the OS Explorer into the project (works with multiple files and empty folders) https://github.com/Jooshua/e4CompareFramework-team2/issues/10
- Workspace Structure Setup (Extensionpoints available to other projects, setting up the initial tree structure) https://github.com/Jooshua/e4CompareFramework-team2/issues/13
- Refactor Codebase to conform to FX Model-View-Controller pattern https://github.com/Jooshua/e4CompareFramework-team2/issues/19

Note that the last bullet point only moves existing logic between files, while refactoring Class and Package names.

### CODE REVIEW

File: de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/CustomTreeCell.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+/**
+ * A tree cell that supports dragging
+ */
+public class CustomTreeCell extends TextFieldTreeCell<FileTreeElement> {
+
+	public CustomTreeCell(WorkspaceFileSystem workspaceFileSystem) {
+		super();
+
+		setOnDragOver(new EventHandler<DragEvent>() {
```
Comment (KamilRosiak) [inline de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/CustomTreeCell.java]: use lambda expression e-> {
}

**Review by KamilRosiak — CHANGES_REQUESTED** [review]
Benutzt den Autoformatter

File: de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/listeners/OpenFileListener.java [hunk 2]
```diff
@@ -1,34 +1,56 @@
...
+			if(!target.isDirectory()) {
+				openFile(target);
+			}
+		}
+	}
+	
+	private void openFile(FileTreeElement target) {
+		if(target != null) {
+			String path = target.getAbsolutePath();
+			String fileExtension = path.substring(path.lastIndexOf(".")+1);
```
Comment (paulo-haas) [inline de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/listeners/OpenFileListener.java]: `getExtension()` is provided by FileTreeElement

File: de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/listeners/OpenFileListener.java [hunk 3]
```diff
@@ -1,34 +1,56 @@
...
+	 * 
+	 * @param eclipseContext
+	 */
+	@Override
+	public void handle(MouseEvent event) {
+		// left mouse button and double click required
+		if(event.getButton().equals(MouseButton.PRIMARY) && event.getClickCount() == 2) {
+			// should not trigger on directories
+			FileTreeElement target = services.rcpSelectionService.getCurrentSelectionFromExplorer();
+			if(!target.isDirectory()) {
```
Comment (paulo-haas) [inline de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/listeners/OpenFileListener.java]: Does this work when `target` equals `null`?

**Review by paulo-haas — CHANGES_REQUESTED** [review]

File: de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/CustomTreeCell.java (hunk 1 above)
Comment (kyshidar) [reply de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/CustomTreeCell.java]: has been updated and tested

File: de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/listeners/OpenFileListener.java (hunk 2 above)
Comment (kyshidar) [reply de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/listeners/OpenFileListener.java]: improvement has been applied

File: de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/listeners/OpenFileListener.java (hunk 3 above)
Comment (kyshidar) [reply de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/listeners/OpenFileListener.java]: now also checks target for null

**Comment by kyshidar** [comment]
the list of changes is long, but aside from the proposed changes everything else was done by the autoformatter and should not require further review

File: de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/ProjectExplorerLabelProvider.java [hunk 4]
```diff
@@ -52,34 +55,36 @@ public boolean isLabelProperty(Object element, String property) {
...
 					image = fileExtensions.get(fileExtension).getIcon(imageService);
 				} else if (fileExtension.equals(E4CStringTable.FILE_ENDING_XML)) {
 					image = imageService.getImage(null, "icons/Explorer_View/items/xml16.png");
 				} else {
-					//default file icon
+					// default file icon
 					image = imageService.getImage(null, "icons/Explorer_View/items/file16.png");
-				};
+				}
+				;
```
Comment (Blum-P) [inline de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/ProjectExplorerLabelProvider.java]: Are you sure to let that semicolon there?

**Review by paulo-haas — APPROVED** [review]

File: de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/ProjectExplorerLabelProvider.java (hunk 4 above)
Comment (Jooshua) [reply de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/ProjectExplorerLabelProvider.java]: Resolved :)

**Review by paulo-haas — APPROVED** [review]

**Pull request merged**